### DISCUSSION TITLE
How to ensure `env_prefix` is respected when using `env_file`?

### DISCUSSION THREAD

**Discussion by ciaransweet** [pydantic/pydantic#10130] [General]
Hey folks,

I have a situation like so:

```python
from pydantic import Field
from pydantic_settings import BaseSettings, SettingsConfigDict

class Config(BaseSettings):
    a_url: str
    model_config = SettingsConfigDict(
        env_file=".env",
        env_prefix="PREFIX_"
    )
```

If I provide `A_URL` as a value in the `.env`, it is populated. If I provide `PREFIX_A_URL`, it is also populated. If I provide **both**, the non-prefixed one is populated.

How do I use `env_prefix` to actually populate my settings with **only** prefixed variables?

I've tried `extra="ignore"` and `extra="forbid"` too and no luck. Is this a quirk of using an `env_file`? I'd rather not implement a custom validator when it feels like these parameters _should_ do that for me?

Appreciate any input people have!

**Comment by hramezani** [comment]
It mentioned in the [docs](https://docs.pydantic.dev/latest/concepts/pydantic_settings/#parsing-environment-variable-values)
```
Pydantic settings loads all the values from dotenv file and passes it to the model, regardless of the model's env_prefix
```

So, `A_URL` will be collected because of the above reason.

**Reply by ciaransweet** [reply]
Okay, so indeed the `env_prefix` is pointless if using a dotenv file

Is there a reason behind that do you know?

I suppose to keep the behaviour, I would have to implement a validator as I mentioned. It's a shame the `extra` parameter isn't used :D

**Reply by hramezani** [reply]
> Okay, so indeed the env_prefix is pointless if using a dotenv file

No, `pydantic-settings` cares about `env_prefix`. as you mentioned in your second case (`PREFIX_A_URL`).

> Is there a reason behind that do you know?

Because of extra config, we had to collect all the values in the dotenv file. consider that you provide `extra='forbid'` and have an extra key in your dotenv file. without collecting all the values we can't raise a `Extra inputs are not permitted` validation error

> I suppose to keep the behaviour, I would have to implement a validator as I mentioned. It's a shame the extra parameter isn't used

`pydantic-settings` cares about extra. in your case, `A_URL` will be collected and passed to the model, that's why you will see always the value of `A_URL`

**Reply by ciaransweet** [reply]
It somewhat makes it redundant though right?

If I *want* only prefixed values, I can't use these parameters as is?

`pydantic-settings` cares about it, but ultimately my settings class is populated with values that don't meet my prefix criteria
